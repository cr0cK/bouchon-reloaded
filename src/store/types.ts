export type State = Record<string, any>

export type ActionId = string

export type Action = () => {
  name: string
  id: ActionId
}

export type Reducer = () => State

export type Reducers = Map<ActionId, Reducer>
