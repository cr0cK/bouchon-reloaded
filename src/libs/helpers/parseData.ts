import * as fs from 'fs'
import { newLogger } from './logger'

/**
 * Parse JSON and verify integrify via a JSON schema.
 *
 * Usage:
 *
 * const { ... } = createCork<...>(
 *   parseData('users-data.json')
 * )
 */
export function parseData<TStore>(dataPathname: string): TStore {
  const logger = newLogger('parseJSONAsStore')

  try {
    // TODO Use AJV to verify JSON integrity

    return JSON.parse(fs.readFileSync(dataPathname, 'utf8')) as TStore
  } catch (err) {
    logger.error(`Can't parse ${dataPathname}, exiting.`, err)
    process.exit(1)
  }
}
