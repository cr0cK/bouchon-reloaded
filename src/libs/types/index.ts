import * as express from 'express'

/** Helpers */

export type Perhaps<T> = T | null | undefined
export type MaybeUndef<T> = T | undefined
export type Maybe<T> = T | null

/** State machine */

export type State = Record<string, any>

export interface Action {
  name: string
  parameters?: object
  bodyParameters?: object
  headerParameters?: object
  queryParameters?: object
}

export type ActionFn<A> = (action: A) => void

export type Reducer<TState extends State, TActionProps> = (
  state: TState,
  action: TActionProps
) => TState

export type Reducers<
  TState extends State,
  TActionEnum extends string,
  TActionsRecord extends Record<TActionEnum, object>
> = Map<keyof TActionsRecord, Reducer<TState, any>>

/** Selectors */

export type Selector<TState, TActionUnion, TSelectorReturn> = (
  state: TState,
  action: TActionUnion
) => TSelectorReturn

export type SelectorFn<TActionUnion, TSelectorReturn> = (
  action: TActionUnion
) => TSelectorReturn

/** Routing */

export type Route<TActionUnion, TSelectorReturn> = {
  method: 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE' | 'HEAD'
  pathname: string
  action: ActionFn<any>
  selector?: SelectorFn<TActionUnion, TSelectorReturn>
  handler?: (
    selectedData: TSelectorReturn,
    action: TActionUnion
  ) => (req: express.Request, res: express.Response) => void
}

export type AnyRoute = Route<any, any>

export type EndPoint = {
  pathname: string
  routes: Array<Route<any, any>>
}

/** Types helpers */

export interface IMockIdRelation {
  mockId: MockId
  parentMockId: MaybeUndef<MockId>
}

// useful to identify generated items
export type MockId = number
