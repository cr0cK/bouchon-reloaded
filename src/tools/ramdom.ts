import { random, range, sampleSize, values } from 'lodash'

/**
 * Return one of the two variables.
 */
export function randomValueBetween<T>(...allValues: T[]): T {
  return allValues[random(0, allValues.length - 1)]
}

/**
 * Return a random value in the range.
 */
export function randomNumberInRange(rangeValues: number[]): number {
  return random(rangeValues[0], rangeValues[1])
}

/**
 * Shortcut for `range()` between 2 bounds.
 */
export function rangeNumberBetween(rangeValues: number[]): number[] {
  return range(0, random(rangeValues[0], rangeValues[1]))
}

/**
 * Return a ramdom value of the enum T.
 */
export function randomValueFromEnum<T>(enumValue: object): T {
  return sampleSize(enumValue, 1).pop() as unknown as T
}

/**
 * Return ramdom values from the enum T.
 */
export function randomValuesFromEnum<T>(enumValue: object): T[] {
  const allValuesLength = values(enumValue).length
  return sampleSize(enumValue, randomNumberInRange([0, allValuesLength - 1]))
}

/**
 * Return a random number between 2 bounds.
 */
export function randomValue(rangeNumbers: number[]) {
  return random(rangeNumbers[0], rangeNumbers[1])
}
