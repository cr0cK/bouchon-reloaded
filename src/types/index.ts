/** State machine */

export type State = Record<string, any>

export interface ActionProps {
  name: string
  parameters?: object
  bodyParameters?: object
  headerParameters?: object
}

export type Reducer<TState extends State, TActionProps> = (
  state: TState,
  actionProps: TActionProps
) => TState

export type Reducers<
  TState extends State,
  TActionEnum extends string,
  TActionsRecord extends Record<TActionEnum, object>
> = Map<keyof TActionsRecord, Reducer<TState, any>>

/** Selectors */

export type SelectorFn<S, AP, R> = (state: S, actionProps: AP) => R

export type Selector<TState, AP, R> = (
  selectorFn: SelectorFn<TState, AP, R>
) => R

/** Routing */

export type Route<TState, AP, R> = {
  method: 'GET' | 'POST' | 'PATH' | 'PUT' | 'DELETE' | 'HEAD'
  pathname: string
  selector: Selector<TState, AP, R>
}
