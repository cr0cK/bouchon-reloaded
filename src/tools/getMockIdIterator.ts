import { MockId } from '../libs/types'

/**
 * Return an iterator that yeilds new MockId.
 */
export function* getMockIdIterator<T extends string>(
  entities: T[],
  forEntity: T
): Iterator<MockId> {
  let firstMockId = (entities.indexOf(forEntity) + 1) * 100000 + 1

  while (true) {
    yield firstMockId
    firstMockId += 1
  }
}
