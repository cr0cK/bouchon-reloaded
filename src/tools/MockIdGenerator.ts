import * as fs from 'fs'
import { sample, sampleSize } from 'lodash'
import { ensureArray } from '../libs/helpers'
import { replacer, reviver } from '../libs/helpers/jsonSerializer'
import { logger } from '../libs/logger'
import { MockId } from '../libs/types'
import { IMockIdRelation, Maybe, MaybeUndef } from '../libs/types'
import { getMockIdIterator } from './getMockIdIterator'
import { rangeNumberBetween } from './ramdom'

export interface IMockIdGeneratorFileFormat<TMockEntityName> {
  allMockIds: Map<TMockEntityName, IMockIdRelation[]>
  metas: Map<string, Map<TMockEntityName, Map<string, any>>>
}

/**
 * Generate mockId to reference mocked data relations.
 */
export class MockIdGenerator<TMockEntityName extends string> {
  private _entityNames: TMockEntityName[] = []

  private _allMockIds: IMockIdGeneratorFileFormat<TMockEntityName>['allMockIds'] =
    new Map()

  private _mockIdIterators: Map<TMockEntityName, Iterator<MockId>> = new Map()

  private _metas: IMockIdGeneratorFileFormat<TMockEntityName>['metas'] =
    new Map()

  /**
   * Set entity names that will be used to generate mockIds.
   */
  constructor(entityNames: TMockEntityName[]) {
    this._entityNames = entityNames
  }

  /**
   * Generate a new id for a collection.
   */
  newMockId(forEntity: TMockEntityName, parentMockId?: MockId): MockId {
    if (!this._allMockIds.has(forEntity)) {
      this._allMockIds.set(forEntity, [])
    }

    if (!this._mockIdIterators.has(forEntity)) {
      this._mockIdIterators.set(
        forEntity,
        getMockIdIterator(this._entityNames, String(forEntity))
      )
    }

    const mockId = this.getIterator(forEntity).next().value

    this.getMockIdRelation(forEntity).push({
      mockId,
      parentMockId
    })

    return mockId
  }

  /**
   * Generate a new list of ids for a collection.
   */
  newListOfMockIds(
    forEntity: TMockEntityName,
    rangeValues: MockId[],
    parentMockId?: MockId
  ): MockId[] {
    return rangeNumberBetween(rangeValues).map(() =>
      this.newMockId(forEntity, parentMockId)
    )
  }

  /**
   * Return the iterator of an entityName.
   */
  getIterator(forEntity: TMockEntityName): Iterator<MockId> {
    const iterator = this._mockIdIterators.get(forEntity)

    if (!iterator) {
      throw new Error('Cant retrieve iterator')
    }

    return iterator
  }

  /**
   * Retrieve a first ID from a collection.
   */
  getFirstMockId(
    forEntity: TMockEntityName,
    withParentMockId?: MockId
  ): MockId {
    const uuids = this.getMockIdRelation(forEntity, withParentMockId).map(
      uuid => uuid.mockId
    )
    return uuids.shift() || -1
  }

  /**
   * Retrieve a list of id from a collection
   */
  getRandomMockIds(
    forEntity: TMockEntityName,
    howMany: MockId,
    withParentMockId?: MockId
  ): MockId[] {
    const mockIds = this.getMockIdRelation(forEntity, withParentMockId).map(
      uuid => uuid.mockId
    )
    return sampleSize(mockIds, howMany)
  }

  /**
   * Retrieve an id from a collection.
   */
  getRandomMockId(
    forEntity: TMockEntityName,
    withParentMockId?: MockId
  ): MockId {
    return this.getRandomMockIds(forEntity, 1, withParentMockId).pop() || -1
  }

  /**
   * Retrieve a list of ids from a collection.
   */
  getListOfMocksIds(
    forEntity: TMockEntityName,
    withParentMockId?: MockId
  ): MockId[] {
    return this.getMockIdRelation(forEntity, withParentMockId).map(
      u => u.mockId
    )
  }

  /**
   * Retrieve the parent of an id from a collection.
   */
  getParentMockId(forEntity: TMockEntityName, mockId: MockId): MockId {
    const mockIdRelation = this.getMockIdRelation(forEntity)
      .filter(u => u.mockId === mockId)
      .pop()

    if (!mockIdRelation) {
      return -1
    }

    return mockIdRelation.parentMockId || -1
  }

  /**
   * Return the collection of an entity.
   */
  getMockIdRelation(
    forEntity: TMockEntityName,
    withParentMockId?: MockId
  ): IMockIdRelation[] {
    if (!this._allMockIds.has(forEntity)) {
      this._allMockIds.set(forEntity, [])
    }

    if (!withParentMockId) {
      return ensureArray(this._allMockIds.get(forEntity))
    }

    return ensureArray(this._allMockIds.get(forEntity)).filter(
      entry => entry.parentMockId === withParentMockId
    )
  }

  /**
   * Return a map from meta data.
   */
  getMetaMap<T>(name: string, forEntity: TMockEntityName): Map<string, T> {
    const map = this._metas.get(name)

    if (!map) {
      return new Map()
    }

    const metaMap = map.get(forEntity)

    return metaMap || new Map()
  }

  /**
   * Return a random entry of the map from meta data.
   */
  getRandomFromMetaMap<T>(name: string, forEntity: TMockEntityName): Maybe<T> {
    const map = this.getMetaMap<T>(name, forEntity)
    const mapKeys = Array.from(map.keys())
    const randomKey = sample(mapKeys)

    if (!randomKey) {
      return null
    }

    return map.get(randomKey) || null
  }

  /**
   * Return a meta data.
   */
  getMeta<T>(
    name: string,
    forEntity: TMockEntityName,
    id: Maybe<MockId>
  ): MaybeUndef<T> {
    if (!id) {
      return
    }

    const metaMap = this.getMetaMap<T>(name, forEntity)
    return metaMap.get(String(id))
  }

  /**
   * Save a meta data.
   */
  setMeta<T>(name: string, forEntity: TMockEntityName, id: MockId, meta: T) {
    const map = this._metas.get(name) || new Map()
    const metas = map.get(forEntity) || new Map()

    metas.set(String(id), meta)
    map.set(forEntity, metas)

    this._metas.set(name, map)
  }

  /**
   * Unshift an uuid in the collection of an entity.
   * Useful for processes that will "consume" a list of UUID
   * (like the deviance routine in the WS server).
   */
  shift(forEntity: TMockEntityName): MaybeUndef<IMockIdRelation> {
    const uuids = this._allMockIds.get(forEntity)
    if (!uuids) {
      return
    }

    return uuids.shift()
  }

  /**
   * Return stringified JSON of all collections.
   */
  stringify(): string {
    const fileContent: IMockIdGeneratorFileFormat<TMockEntityName> = {
      allMockIds: this._allMockIds,
      metas: this._metas
    }

    return JSON.stringify(fileContent, replacer, 2)
  }

  /**
   * Retrieve all collections from a JSON string.
   */
  loadFromFile(filePath: string): void {
    try {
      const content = fs.readFileSync(filePath, 'utf-8')
      const parsedContent = JSON.parse(
        content,
        reviver
      ) as IMockIdGeneratorFileFormat<TMockEntityName>

      this._allMockIds = new Map(parsedContent.allMockIds)
      this._metas = new Map(parsedContent.metas)
    } catch (err) {
      if (err instanceof Error) {
        logger.error(err)

        if (err.stack) {
          logger.debug(err.stack)
        }
      }
    }
  }
}
