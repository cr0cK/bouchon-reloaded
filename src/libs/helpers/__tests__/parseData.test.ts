import { MockIdGenerator } from '../../../tools/MockIdGenerator'
import { MockId } from '../../types'
import { stringifyData } from '../parseData'

enum MockEntity {
  user = 'user'
}

interface IUser {
  name: string
  email: string
}

describe('parseData', () => {
  describe('stringifyData', () => {
    let data: { users: Map<MockId, IUser> }

    beforeEach(() => {
      const mockIdsGenerator = new MockIdGenerator<MockEntity>(
        Object.values(MockEntity)
      )

      data = {
        users: new Map([
          [
            mockIdsGenerator.newMockId(MockEntity.user),
            {
              name: 'Bob',
              email: 'bob@aol.com'
            }
          ],
          [
            mockIdsGenerator.newMockId(MockEntity.user),
            {
              name: 'Alice',
              email: 'alice@aol.com'
            }
          ]
        ])
      }
    })

    it('should stringify data', () => {
      const str = stringifyData(data)
      expect(str).toMatchSnapshot()
    })
  })
})
