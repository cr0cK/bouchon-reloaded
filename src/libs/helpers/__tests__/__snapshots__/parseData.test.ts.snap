// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseData stringifyData should stringify data 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        100001,
        {
          "name": "Bob",
          "email": "bob@aol.com"
        }
      ],
      [
        100002,
        {
          "name": "Alice",
          "email": "alice@aol.com"
        }
      ]
    ]
  }
}"
`;
